src/components/terrain-panel.tsx

```diff
@@ -10,230 +10,277 @@ import { droughtLabels, DroughtLevel, terrainLabels, TerrainType, Vegetation, ve
-      this.selectedZone = newZone;
+    if (newZone !== selectedZone) {
+      setSelectedZone(newZone);
     }
     log("TerrainPanelZoneChanged", { zone: newZone });
   };
 
-  public showNextPanel = () => {
-    this.setState({ currentPanel: 2 });
+  const showNextPanel = () => {
+    if (currentPanel === 0) {
+      // Delay zones count change until user clicks "Next" button, as it might lead to some unavoidable user setup loss.
+      applyZonesCountChange();
+    }
+    setCurrentPanel(val => val + 1);
     log("TerrainPanelNextButtonClicked");
   };
 
-  public showPreviousPanel = () => {
-    this.setState({ currentPanel: 1 });
+  const showPreviousPanel = () => {
+    setCurrentPanel(val => val - 1);
     log("TerrainPanelPreviousButtonClicked");
   };
 
-  public handleTerrainTypeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-    const { simulation } = this.stores;
+  const handleTerrainTypeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     const newTerrainType = parseInt(event.target.value, 10) as TerrainType;
-    const currentZone = simulation.zones[this.selectedZone];
-    const logData: any = { zone: this.selectedZone };
+    const currentZone = zones[selectedZone];
+    const logData: any = { zone: selectedZone };
     if (currentZone.terrainType !== newTerrainType) {
       // Switching to Mountain terrain changes land type / vegetation options
       // but keeping the min / max options the same for each range helps with slider rendering.
       // Accommodate this by manual adjustment of land types when switching to-from mountain
       if (newTerrainType !== TerrainType.Mountains && currentZone.vegetation === Vegetation.ForestWithSuppression) {
         // switching from Mountains with large forests to lower land type, reduce forest size
-        simulation.updateZoneVegetation(this.selectedZone, Vegetation.Forest);
+        currentZone.vegetation = Vegetation.Forest;
         logData.vegetation = vegetationLabels[Vegetation.Forest];
       } else if (newTerrainType === TerrainType.Mountains && currentZone.vegetation === Vegetation.Grass) {
         // no grass allowed on mountains, switch to shrubs
-        simulation.updateZoneVegetation(this.selectedZone, Vegetation.Shrub);
+        currentZone.vegetation = Vegetation.Shrub;
         logData.vegetation = vegetationLabels[Vegetation.Shrub];
       }
-      simulation.updateZoneTerrain(this.selectedZone, newTerrainType);
+      currentZone.terrainType = newTerrainType;
       logData.terrain = terrainLabels[newTerrainType];
       log("ZoneUpdated", logData);
     }
   };
 
-  public handleVegetationChange = (event: Event, value: number) => {
-    const { simulation } = this.stores;
-    const zone = simulation.zones[this.selectedZone];
+  const handleVegetationChange = (event: Event, value: number) => {
     if (zone.vegetation !== value) {
-      simulation.updateZoneVegetation(this.selectedZone, value);
+      zone.vegetation = value;
```

Comment:
It seems kind of non-standard to directly set a `useState` variable instead of using its setter.  I see in the `applyAndClose` handler you are then using the simulation `setZones` method to finalize the updates.  I wonder if it would be clearer/cleaner to use a ref for the zones/zone instead since you are not using its setter?